test/api_bridge_test.rb

```diff
@@ -0,0 +1,165 @@
+    # ampersand!
+    param = @api.encode_param("f[]=repository|Archives & Special Collections")
+    assert_equal "f[]=repository%7CArchives+%26+Special+Collections", param
+    # numbers
+    param = @api.encode_param("facet[]=date.year|1867")
+    assert_equal "facet[]=date.year%7C1867", param
+  end
+
+  # get_item_by_id
+
+  test "is_url?" do
+    # definitely not
+    assert_not @api.is_url?("some string")
+    # sort of but not quite!
+    assert_not @api.is_url?("test.unl.edu")
+    # http
+    assert @api.is_url?("http://test.unl.edu")
+    # https
+    assert @api.is_url?("https://test.unl.edu")
+    # port and path
+    assert @api.is_url?("http://localhost:3000/api/v1/items")
+  end
+
+  test "params_to_hash" do
+    # make sure this hash is not altered and returns only as string keys
+    original = { "a" => 1, b: 2 }
+    cloned = @api.params_to_hash(original)
+    assert_includes cloned, "b"
+    assert_not_includes cloned, :b
+    cloned.delete("a")
+    assert_includes original, "a"
+    # TODO ActionController::Parameters object next!
+    original = ActionController::Parameters.new({ "a" => 1, b: 2 })
+    cloned = @api.params_to_hash(original)
+    assert_includes cloned, "b"
+    assert_not_includes cloned, :b
+    cloned.delete("a")
+    assert_includes original, "a"
+  end
+
+  test "prepare_options" do
+    # make sure rows does not override num when also sent in
+    # (@api created with num)
+    opts = @api.prepare_options({ "rows" => 20 })
+    assert_not_includes opts, "rows"
+    assert_equal 5, opts["num"]
+    # TODO figure out if we want rows to override num even though
+    # there is a default num set...will need to figure something out
+    # to determine if num was the default or passed in....??
+    # api = ApiBridge::Query.new("https://test.unl.edu/v1", [], {})
+    # opts = api.prepare_options({ "rows" => 20 })
+    # assert_not_includes opts, "rows"
+    # assert_equal 20, opts["num"]
+
+    # make sure to remove rails stuff
+    req_opts = ActionController::Parameters.new({
+      controller: "items", action: "index", utf8: "✓", q: "water"
+    })
+    res_opts = { "num"=>5, "sort"=>"title|asc", "facet"=>["creator.name", "date.year", "subcategory"], "q"=>"water", "start"=>0 }
+    assert_equal res_opts, @api.prepare_options(req_opts)
+  end
+
+  test "query" do
+    skip("skipping integration testing")
+  end
+
+  test "remove_rows" do
+    # TODO see commentary in prepare_options above
+    # we should perhaps prefer rows to num for backwards compatibility
+    # if they are sending it for a specific request
+  end
+
+  test "send_request" do
+    skip "skipping integration testing"
+  end
+
+  test "set_page" do
+    # no page means page 1!
+    assert_equal 1, @api.set_page(nil)
+    # page can't be zero or negative, sorry
+    assert_equal 1, @api.set_page("0")
+    assert_equal 1, @api.set_page("00")
+    assert_equal 1, @api.set_page("-1")
+    # page has to be some type of number
```

Comment:
The code was modified to handle leading zeros so there should probably be a test like:
```ruby
# leading zeros not allowed
assert_equal 1, @api.set_page("09")
```

Reply:
👍